spec/controllers/two_factor_authentication/piv_cac_verification_controller_spec.rb

```diff
@@ -0,0 +1,187 @@
+        UpdateUser.new(
+          user: subject.current_user,
+          attributes: { second_factor_attempts_count: 1 }
+        ).call
+
+        get :show, params: { token:  'good-token' }
+
+        expect(subject.current_user.reload.second_factor_attempts_count).to eq 0
+      end
+
+      it 'tracks the valid authentication event' do
+        stub_analytics
+        attributes = {
+          success: true,
+          errors: {},
+          context: 'authentication',
+          multi_factor_auth_method: 'piv_cac',
+        }
+        expect(@analytics).to receive(:track_event).with(Analytics::MULTI_FACTOR_AUTH, attributes)
+
+        get :show, params: { token:  'good-token' }
+      end
+    end
+
+    context 'when the user presents an invalid piv/cac' do
+      before do
+        sign_in_before_2fa(user)
+
+        get :show, params: { token: 'bad-token' }
+      end
+
+      it 'increments second_factor_attempts_count' do
+        expect(subject.current_user.reload.second_factor_attempts_count).to eq 1
+      end
+
+      it 're-renders the piv/cac entry screen' do
+        expect(response).to render_template(:show)
+      end
+
+      it 'displays flash error message' do
+        expect(flash[:error]).to eq t('devise.two_factor_authentication.invalid_piv_cac')
+      end
+    end
+
+    context 'when the user has reached the max number of piv/cac attempts' do
+      it 'tracks the event' do
+        allow_any_instance_of(User).to receive(:max_login_attempts?).and_return(true)
+        sign_in_before_2fa(user)
+
+        stub_analytics
+
+        attributes = {
+          success: false,
+          errors: {},
+          context: 'authentication',
+          multi_factor_auth_method: 'piv_cac',
+        }
+
+        expect(@analytics).to receive(:track_event).with(Analytics::MULTI_FACTOR_AUTH, attributes)
+        expect(@analytics).to receive(:track_event).with(Analytics::MULTI_FACTOR_AUTH_MAX_ATTEMPTS)
+
+        get :show, params: { token: 'bad-token' }
+      end
+    end
+
+    context 'when the user lockout period expires' do
+      before(:each) do
+        sign_in_before_2fa(user)
+      end
+
+      let(:lockout_period) { Figaro.env.lockout_period_in_minutes.to_i.minutes }
+
+      let(:user) do
+        create(:user, :signed_up, :with_piv_or_cac,
+          second_factor_locked_at: Time.zone.now - lockout_period - 1.second,
+          second_factor_attempts_count: 3
+        )
+      end
+
+      describe 'when user submits an incorrect piv/cac' do
+        before(:each) do
+          get :show, params: { token: 'bad-token' }
+        end
+
+        it 'resets attempts count' do
+          expect(subject.current_user.reload.second_factor_attempts_count).to eq 1
+        end
+
+        it 'resets second_factor_locked_at' do
+          expect(subject.current_user.reload.second_factor_locked_at).to eq nil
+        end
+      end
+
+      describe 'when user submits a valid TOTP' do
```

Comment:
s/TOTP/(PIV/CAC)?

Reply:
yup - I thought I got all those -- I was copying one of the other 2fa specs and changing it to work for piv/cac to make sure I covered the same situations that applied to piv/cac.